File: reference/FlaMEVB/clsPainter.vb

Public Class clsPainter
    Public Class clsTerrain
        Public Num As Integer
        Public Name As String

        Public Tiles As New clsTileList
    End Class
    Public Terrains() As clsTerrain
    Public TerrainCount As Integer
    Public Class clsTileList
        Public Structure sTileOrientationChance
            Public TextureNum As Integer
            Public Direction As sTileDirection
            Public Chance As UInteger
        End Structure
        Public Tiles() As sTileOrientationChance
        Public TileCount As Integer
        Public TileChanceTotal As Integer

        Public Sub Tile_Add(TileNum As Integer, TileOutwardOrientation As sTileDirection, Chance As UInteger)

            ReDim Preserve Tiles(TileCount)
            Tiles(TileCount).TextureNum = TileNum
            Tiles(TileCount).Direction = TileOutwardOrientation
            Tiles(TileCount).Chance = Chance
            TileCount += 1

            TileChanceTotal += CInt(Chance)
        End Sub

        Public Sub Tile_Remove(Num As Integer)

            TileChanceTotal -= CInt(Tiles(Num).Chance)

            TileCount -= 1
            If Num <> TileCount Then
                Tiles(Num) = Tiles(TileCount)
            End If
            ReDim Preserve Tiles(TileCount - 1)
        End Sub

        Public Function GetRandom() As sTileOrientationChance
            Dim ReturnResult As sTileOrientationChance
            Dim A As Integer
            Dim intRandom As Integer
            Dim Total As Integer

            intRandom = CInt(Int(Rnd() * TileChanceTotal))
            For A = 0 To TileCount - 1
                Total += CInt(Tiles(A).Chance)
                If intRandom < Total Then
                    Exit For
                End If
            Next
            If A = TileCount Then
                ReturnResult.TextureNum = -1
                ReturnResult.Direction = TileDirection_None
            Else
                ReturnResult = Tiles(A)
            End If
            Return ReturnResult
        End Function
    End Class
    Public Class clsTransition_Brush
        Public Name As String
        Public Terrain_Inner As clsTerrain
        Public Terrain_Outer As clsTerrain
        Public Tiles_Straight As New clsTileList
        Public Tiles_Corner_In As New clsTileList
        Public Tiles_Corner_Out As New clsTileList
    End Class
    Public TransitionBrushes() As clsTransition_Brush
    Public TransitionBrushCount As Integer
    Public Class clsCliff_Brush
        Public Name As String
        Public Terrain_Inner As clsTerrain
        Public Terrain_Outer As clsTerrain
        Public Tiles_Straight As New clsTileList
        Public Tiles_Corner_In As New clsTileList
        Public Tiles_Corner_Out As New clsTileList
    End Class
    Public CliffBrushes() As clsCliff_Brush
    Public CliffBrushCount As Integer

    Public Class clsRoad
        Public Num As Integer
        Public Name As String
    End Class
    Public Roads() As clsRoad
    Public RoadCount As Integer
    Public Class clsRoad_Brush
        Public Road As clsRoad
        Public Terrain As clsTerrain
        Public Tile_Straight As New clsTileList
        Public Tile_Corner_In As New clsTileList
        Public Tile_TIntersection As New clsTileList
        Public Tile_CrossIntersection As New clsTileList
        Public Tile_End As New clsTileList
    End Class
    Public RoadBrushes() As clsRoad_Brush
    Public RoadBrushCount As Integer

    Public Sub TransitionBrush_Add(NewBrush As clsTransition_Brush)

        ReDim Preserve TransitionBrushes(TransitionBrushCount)
        TransitionBrushes(TransitionBrushCount) = NewBrush
        TransitionBrushCount += 1
    End Sub

    Public Sub TransitionBrush_Remove(Num As Integer)

        TransitionBrushCount -= 1
        If Num <> TransitionBrushCount Then
            TransitionBrushes(Num) = TransitionBrushes(TransitionBrushCount)
        End If
        ReDim Preserve TransitionBrushes(TransitionBrushCount - 1)
    End Sub

    Public Sub CliffBrush_Add(NewBrush As clsCliff_Brush)

        ReDim Preserve CliffBrushes(CliffBrushCount)
        CliffBrushes(CliffBrushCount) = NewBrush
        CliffBrushCount += 1
    End Sub

    Public Sub CliffBrush_Remove(Num As Integer)

        CliffBrushCount -= 1
        If Num <> CliffBrushCount Then
            CliffBrushes(Num) = CliffBrushes(CliffBrushCount)
        End If
        ReDim Preserve CliffBrushes(CliffBrushCount - 1)
    End Sub

    Public Sub Terrain_Add(NewTerrain As clsTerrain)

        NewTerrain.Num = TerrainCount
        ReDim Preserve Terrains(TerrainCount)
        Terrains(TerrainCount) = NewTerrain
        TerrainCount += 1
    End Sub

    Public Sub Terrain_Remove(Num As Integer)

        Terrains(Num).Num = -1
        TerrainCount -= 1
        If Num <> TerrainCount Then
            Terrains(Num) = Terrains(TerrainCount)
            Terrains(Num).Num = Num
        End If
        ReDim Preserve Terrains(TerrainCount - 1)
    End Sub

    Public Sub RoadBrush_Add(NewRoadBrush As clsRoad_Brush)

        ReDim Preserve RoadBrushes(RoadBrushCount)
        RoadBrushes(RoadBrushCount) = NewRoadBrush
        RoadBrushCount += 1
    End Sub

    Public Sub RoadBrush_Remove(Num As Integer)

        RoadBrushCount -= 1
        If Num <> RoadBrushCount Then
            RoadBrushes(Num) = RoadBrushes(RoadBrushCount)
        End If
        ReDim Preserve RoadBrushes(RoadBrushCount - 1)
    End Sub

    Public Sub Road_Add(NewRoad As clsRoad)

        NewRoad.Num = RoadCount
        ReDim Preserve Roads(RoadCount)
        Roads(RoadCount) = NewRoad
        RoadCount += 1
    End Sub

    Public Sub Road_Remove(Num As Integer)

        Roads(Num).Num = -1
        RoadCount -= 1
        If Num <> RoadCount Then
            Roads(Num) = Roads(RoadCount)
            Roads(Num).Num = Num
        End If
        ReDim Preserve Roads(RoadCount - 1)
    End Sub
End Class

File: source/SharpFlame/clsPainter.cs
using System;
using Microsoft.VisualBasic;

namespace SharpFlame
{
    public class clsPainter
    {
        public class clsTerrain
        {
            public int Num;
            public string Name;

            public clsTileList Tiles = new clsTileList();
        }

        public clsTerrain[] Terrains;
        public int TerrainCount;

        public class clsTileList
        {
            public struct sTileOrientationChance
            {
                public int TextureNum;
                public TileOrientation.sTileDirection Direction;
                public UInt32 Chance;
            }

            public sTileOrientationChance[] Tiles;
            public int TileCount;
            public int TileChanceTotal;

            public void Tile_Add(int TileNum, TileOrientation.sTileDirection TileOutwardOrientation, UInt32 Chance)
            {
                Array.Resize(ref Tiles, TileCount + 1);
                Tiles[TileCount].TextureNum = TileNum;
                Tiles[TileCount].Direction = TileOutwardOrientation;
                Tiles[TileCount].Chance = Chance;
                TileCount++;

                TileChanceTotal += System.Convert.ToInt32(Chance);
            }

            public void Tile_Remove(int Num)
            {
                TileChanceTotal -= System.Convert.ToInt32(Tiles[Num].Chance);

                TileCount--;
                if ( Num != TileCount )
                {
                    Tiles[Num] = Tiles[TileCount];
                }
                Array.Resize(ref Tiles, TileCount);
            }

            public sTileOrientationChance GetRandom()
            {
                sTileOrientationChance ReturnResult = new sTileOrientationChance();
                int A = 0;
                int intRandom = 0;
                int Total = 0;

                intRandom = (int)(Conversion.Int(VBMath.Rnd() * TileChanceTotal));
                for ( A = 0; A <= TileCount - 1; A++ )
                {
                    Total += System.Convert.ToInt32(Tiles[A].Chance);
                    if ( intRandom < Total )
                    {
                        break;
                    }
                }
                if ( A == TileCount )
                {
                    ReturnResult.TextureNum = -1;
                    ReturnResult.Direction = TileOrientation.TileDirection_None;
                }
                else
                {
                    ReturnResult = Tiles[A];
                }
                return ReturnResult;
            }
        }

        public class clsTransition_Brush
        {
            public string Name;
            public clsTerrain Terrain_Inner;
            public clsTerrain Terrain_Outer;
            public clsTileList Tiles_Straight = new clsTileList();
            public clsTileList Tiles_Corner_In = new clsTileList();
            public clsTileList Tiles_Corner_Out = new clsTileList();
        }

        public clsTransition_Brush[] TransitionBrushes;
        public int TransitionBrushCount;

        public class clsCliff_Brush
        {
            public string Name;
            public clsTerrain Terrain_Inner;
            public clsTerrain Terrain_Outer;
            public clsTileList Tiles_Straight = new clsTileList();
            public clsTileList Tiles_Corner_In = new clsTileList();
            public clsTileList Tiles_Corner_Out = new clsTileList();
        }

        public clsCliff_Brush[] CliffBrushes;
        public int CliffBrushCount;

        public class clsRoad
        {
            public int Num;
            public string Name;
        }

        public clsRoad[] Roads;
        public int RoadCount;

        public class clsRoad_Brush
        {
            public clsRoad Road;
            public clsTerrain Terrain;
            public clsTileList Tile_Straight = new clsTileList();
            public clsTileList Tile_Corner_In = new clsTileList();
            public clsTileList Tile_TIntersection = new clsTileList();
            public clsTileList Tile_CrossIntersection = new clsTileList();
            public clsTileList Tile_End = new clsTileList();
        }

        public clsRoad_Brush[] RoadBrushes;
        public int RoadBrushCount;

        public void TransitionBrush_Add(clsTransition_Brush NewBrush)
        {
            Array.Resize(ref TransitionBrushes, TransitionBrushCount + 1);
            TransitionBrushes[TransitionBrushCount] = NewBrush;
            TransitionBrushCount++;
        }

        public void TransitionBrush_Remove(int Num)
        {
            TransitionBrushCount--;
            if ( Num != TransitionBrushCount )
            {
                TransitionBrushes[Num] = TransitionBrushes[TransitionBrushCount];
            }
            Array.Resize(ref TransitionBrushes, TransitionBrushCount);
        }

        public void CliffBrush_Add(clsCliff_Brush NewBrush)
        {
            Array.Resize(ref CliffBrushes, CliffBrushCount + 1);
            CliffBrushes[CliffBrushCount] = NewBrush;
            CliffBrushCount++;
        }

        public void CliffBrush_Remove(int Num)
        {
            CliffBrushCount--;
            if ( Num != CliffBrushCount )
            {
                CliffBrushes[Num] = CliffBrushes[CliffBrushCount];
            }
            Array.Resize(ref CliffBrushes, CliffBrushCount);
        }

        public void Terrain_Add(clsTerrain NewTerrain)
        {
            NewTerrain.Num = TerrainCount;
            Array.Resize(ref Terrains, TerrainCount + 1);
            Terrains[TerrainCount] = NewTerrain;
            TerrainCount++;
        }

        public void Terrain_Remove(int Num)
        {
            Terrains[Num].Num = -1;
            TerrainCount--;
            if ( Num != TerrainCount )
            {
                Terrains[Num] = Terrains[TerrainCount];
                Terrains[Num].Num = Num;
            }
            Array.Resize(ref Terrains, TerrainCount);
        }

        public void RoadBrush_Add(clsRoad_Brush NewRoadBrush)
        {
            Array.Resize(ref RoadBrushes, RoadBrushCount + 1);
            RoadBrushes[RoadBrushCount] = NewRoadBrush;
            RoadBrushCount++;
        }

        public void RoadBrush_Remove(int Num)
        {
            RoadBrushCount--;
            if ( Num != RoadBrushCount )
            {
                RoadBrushes[Num] = RoadBrushes[RoadBrushCount];
            }
            Array.Resize(ref RoadBrushes, RoadBrushCount);
        }

        public void Road_Add(clsRoad NewRoad)
        {
            NewRoad.Num = RoadCount;
            Array.Resize(ref Roads, RoadCount + 1);
            Roads[RoadCount] = NewRoad;
            RoadCount++;
        }

        public void Road_Remove(int Num)
        {
            Roads[Num].Num = -1;
            RoadCount--;
            if ( Num != RoadCount )
            {
                Roads[Num] = Roads[RoadCount];
                Roads[Num].Num = Num;
            }
            Array.Resize(ref Roads, RoadCount);
        }
    }
}